### PR TITLE
FD to CKAN conversions: Add tests and their implementation

### PR DESCRIPTION
**Pull request by sglavoie**
**TL;DR**: Add all tests discussed in #21, implement them and refactor classes into simple functions as discussed.

* For a granular view of the PR, each commit should be easy to follow. Otherwise, there would be many merge conflicts to deal with, so this should in the end speed you up :slightly_smiling_face:.

### CODE REVIEW

**Comment by rufuspollock** [comment]
My big request for future would be *smaller* PRs 😄 - you can still do a bunch of work and just have multiple PRs i think

**Comment by sglavoie** [comment]
**TL;DR**: Ok, I will do that. :+1: 

I was worried that there would be many merge conflicts as I was refactoring/improving tests along the way and if you needed to merge this early during your day, it would have been stuck in an "unmergeable" state for a while.

I will do my best to work on different sections of the files in each PR to avoid breaking changes.

File: tests/test_frictionless_to_ckan.py [hunk 1]
```diff
@@ -1,30 +1,305 @@
...
+            'extras': {
+                'title_cn': u'國內生產總值',
+                'years': [2015, 2016],
+                'last_year': 2016,
+                'location': {'country': 'China'}
+            }
+        }
+        out = converter.resource(indict)
+        exp = {
+            'extras': [
```
Comment (rufuspollock) [inline tests/test_frictionless_to_ckan.py:18]: This is incorrect. Resources in CKAN do *not* have an extras dict - they just put the keys straight on e.g. 

https://demo.ckan.org/api/3/action/resource_show?id=2feca3fa-b1da-45d8-b14b-7ee174f4623c

```
"result": {
"zip_path": "https://storage.googleapis.com/datopian-nhs/DATOPIAN_DATA/DPI_DETAIL_PRESCRIBING_201712.zip",
"cache_last_updated": null,
"zip_bytes": "514060788",
"package_id": "bb5a18dc-e332-4c6a-b3c7-db6b841aa495",
"datastore_active": false,
"id": "2feca3fa-b1da-45d8-b14b-7ee174f4623c",
"size": null,
"bq_table_id": "bigquerytest-271707.nhs_production.DPI_DETAIL_PRESCRIBING_201712",
"title": "DPI DETAIL PRESCRIBING 201712",
"state": "active",
"bq_table_name": "DPI_DETAIL_PRESCRIBING_201712",
"hash": "",
"description": "",
"format": "CSV",
"mediatype": "text/csv",
"mimetype_inner": null,
"url_type": null,
"zip_name": "DPI_DETAIL_PRESCRIBING_201712.zip",
"mimetype": null,
"cache_url": null,
"zip_title": "DPI DETAIL PRESCRIBING 201712.zip",
"name": "DPI_DETAIL_PRESCRIBING_201712",
"created": "2020-03-31T21:57:50.386342",
"url": "https://storage.googleapis.com/datopian-nhs/csv/DPI_DETAIL_PRESCRIBING_201712.csv",
"bytes": "6442270750",
"last_modified": null,
"position": 47,
"revision_id": "b99d283f-8711-400c-8a12-9b9f1b2409b0",
"zip_format": "zip",
"resource_type": null
}
```

File: tests/test_frictionless_to_ckan.py [hunk 2]
```diff
@@ -1,30 +1,305 @@
...
+                {'key': 'location', 'value': '{"country": "China"}'},
+                {'key': 'title_cn', 'value': u'國內生產總值'},
+                {'key': 'years', 'value': '[2015, 2016]'}
+            ]
+        }
+        out['extras'] = sorted(out['extras'], key = lambda i: i['key'])
+        assert out['extras'] == exp['extras']
 
-converter = frictionless_to_ckan.FrictionlessToCKAN()
+    def test_name_is_used_if_theres_no_title(self):
```
Comment (rufuspollock) [inline tests/test_frictionless_to_ckan.py:28]: This is the kind of magic we don't want to have.

File: tests/test_frictionless_to_ckan.py [hunk 3]
```diff
@@ -1,30 +1,305 @@
...
+        indict = {'name': 'gdp'}
+        out = converter.resource(indict)
+        assert out.get('name') == indict['name']
 
+    def test_resource_title_is_used_as_name(self):
```
Comment (rufuspollock) [inline tests/test_frictionless_to_ckan.py:33]: Again this is magic.

File: tests/test_frictionless_to_ckan.py [hunk 4]
```diff
@@ -1,30 +1,305 @@
...
+        }
 
-class TestResourceConversion:
-    pass
+        out = converter.resource(indict)
+        assert out.get('name') == indict['title']
+
+    def test_path_to_url(self):
+        # Test remote path
+        indict = {'path': 'http://www.somewhere.com/data.csv'}
```
Comment (rufuspollock) [inline tests/test_frictionless_to_ckan.py:44]: please lay out like:

```
indict = {
     ...
}
```

it's easier to read and more consistent if we add other values.

File: tests/test_frictionless_to_ckan.py [hunk 5]
```diff
@@ -1,30 +1,305 @@
...
+        schema = {
+            'fields': [
+                {'name': 'id', 'type': 'integer'},
+                {'name': 'title', 'type': 'string'},
+            ]
+        }
+        indict = {
+            'schema': schema
+        }
+        out = converter.resource(indict)
```
Comment (rufuspollock) [inline tests/test_frictionless_to_ckan.py:77]: this test is wrong - we  should pass  through but with value jsonified ...

**Pull request merged**